### PR TITLE
chore[test]: add macos to test matrix

### PR DESCRIPTION
**Pull request by DanielSchiavini**
### What I did
- Expanded the matrix to include macos and windows tests
- Removed unnecessary lines from the matrix definition

### How I did it
- All default values must be in the top-level matrix definition
- The `includes` section should contain alternative runs - these will not have all default values defined
  - Therefore reading from the matrix requires passing default values again
  - However we can still delete a lot of repetition

### How to verify it
- Tests should pass. We should have 2 extra jobs and the existing ones should keep on existing.

### Commit message

```
- Expanded the matrix to include macos and windows tests
- Refactor the matrix and removed unnecessary lines
- Remove now-redundant windows job

Note that all because of the way github creates matrix jobs, default
values must be in the top-level matrix definition. The includes section
should contain alternative runs - these will not have all default values
defined, therefore reading from the matrix requires passing default
values again.
```

### Description for the changelog
N/A

### Cute Animal Picture
image

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -75,97 +76,79 @@ jobs:
...
-            opt-mode: none
+          - opt-mode: none
             debug: false
-            evm-version: shanghai
-            evm-backend: revm
             experimental-codegen: true
 
+          # os-specific rules
+          - os: windows
+            opt-mode: none
```
Comment (charles-cooper) [inline .github/workflows/test.yml:117]: i don't think we should override the default -- as much as possible, we should try to vary only on a single axis. (like we want to test the most common use case, which will be running with default optimization level)

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -75,97 +76,50 @@ jobs:
...
+          - evm-version: london
+          - evm-version: paris
+          - evm-version: cancun
 
           # py-evm rules
-          - python-version: ["3.11", "311"]
-            debug: false
-            opt-mode: codesize
-            evm-version: london
+          - evm-version: london
```
Comment (charles-cooper) [inline .github/workflows/test.yml:84]: honestly, not sure why we should test py-evm + london. if we are going to have two py-evm runs, it should be cancun + shanghai

Comment (DanielSchiavini) [reply .github/workflows/test.yml:84]:
```suggestion
          - evm-version: shanghai
```

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -75,97 +76,50 @@ jobs:
...
             evm-backend: py-evm
```
Comment (DanielSchiavini) [inline .github/workflows/test.yml:85]:
```suggestion
          - evm-backend: py-evm
```

**Comment by charles-cooper** [comment]
let's put the OS in the front of the job name. we can also drop it in the default case, we only need to distinguish for windows + macos

**Review by charles-cooper — APPROVED** [review]
looks good, thanks!

**Pull request merged**